prov/mrail:  Implement message striping using Rendezvous and RMA

j-xiong opened the pull request:
Implement a Rendezvous protocol that deliver large size messages via RMA
reads. This naturally enables message striping because RMA operations are
always striped across all the rails.
    
Future enhancement will add tuning knobs to the RMA striping algorithm
to improve the performance of medium size messages.
    
Also consolidate two mostly identical common send functions into one.

Signed-off-by: Jianxin Xiong <jianxin.xiong@intel.com>

j-xiong commented:
This is targeted for post 1.8 release. Don't merge before the 1.8 branch is created.

shefty commented on prov/mrail/src/mrail.h:
```diff
@@ -128,16 +128,41 @@ mrail_match_recv_handle_unexp(struct mrail_recv_queue *recv_queue, uint64_t tag,
+#define MRAIL_PROTO_RNDV	1
+
+#define MRAIL_RNDV_REQ		0
+#define MRAIL_RNDV_ACK		1
```
nit: I think these could be defined using enums

j-xiong replied:
That can be done.

shefty commented on prov/mrail/src/mrail_cq.c:
```diff
@@ -51,6 +82,143 @@ int mrail_cq_write_recv_comp(struct mrail_ep *mrail_ep, struct mrail_hdr *hdr,
+				"Cannot send rndv ack: %s\n", fi_strerror(-ret));
+			assert(0);
+		}
+	} while (ret == -FI_EAGAIN);
```
What keeps this from spinning forever?  Does mrail_send_rndv_ack ensure progress?

j-xiong replied:
Nothing. I'll see what can be done here.

shefty commented on prov/mrail/src/mrail_cq.c:
```diff
@@ -51,6 +82,143 @@ int mrail_cq_write_recv_comp(struct mrail_ep *mrail_ep, struct mrail_hdr *hdr,
+	if (ret) {
+		FI_WARN(&mrail_prov, FI_LOG_CQ,
+			"Unable to discard buffered recv\n");
+		retv = ret;
```
use
```
retv = retv ? retv : ret;
```
?

j-xiong replied:
That's probably more desirable since the first error code is more important.

shefty commented on prov/mrail/src/mrail_cq.c:
```diff
@@ -51,6 +82,143 @@ int mrail_cq_write_recv_comp(struct mrail_ep *mrail_ep, struct mrail_hdr *hdr,
+	if (ret) {
+		FI_WARN(&mrail_prov, FI_LOG_CQ,
+			"Unable to discard buffered recv\n");
+		retv = ret;
```
same comment as above

shefty commented on prov/mrail/src/mrail.h:
```diff
@@ -128,16 +128,41 @@ mrail_match_recv_handle_unexp(struct mrail_recv_queue *recv_queue, uint64_t tag,
+	size_t			len;
+	size_t			count;
+	size_t			mr_count;
+	struct fi_rma_iov	rma_iov[MRAIL_IOV_LIMIT];
```
Does the value of MRAIL_IOV_LIMIT restrict the number of rails that can be used?  (I'm okay with that, just checking.)

j-xiong replied:
No. It only affects the buffer layout. Payload is not divided based on iov count.

j-xiong commented:
Updated the PR to address all previous comments. Changed mrail_send_rndv_ack into a blocking call that would make progress internally and wouldn't return -FI_EAGAIN.

shefty commented:
Everything looks fine to me.  The series needs to be rebased though.

j-xiong commented:
Rebased and rechecked with my quick mpi test.
